### ISSUE TITLE
Decouple renderer

### ISSUE THREAD

**Issue by bigopon** [aurelia/aurelia#182]
At the moment, it seems to me `Renderer` has very weak connection with the rest of runtime, and only strongly coupled with instructions list. It seems to make sense to me that we make those instructions at externally configurable, together with a plug-able renderer for those instructions to depend on. This way, we can have default combo of compiler and renderer for web platform. To target different platform, we can provide different combo instead of our current structure. This requires some changes in the way how `runtime` understands the basic behavior, or we can even make all behavior externally configurable, like .. there is no such thing as custom element. It's a renderer thing.
@EisenbergEffect @fkleuver

**Comment by fkleuver** [comment]
Is this different from the `@renderStrategy` decorator I added a while ago?

**Comment by bigopon** [comment]
I think it could be more than that. By separating out the renderer and its infrastructure, we can achieve more than render strategy. Atm there is a fixed set of core instructions, but those instructions are relevant to only renderer and a few more. Maybe should be configurable in bindingcommand resource way instead of patching the prototype way

**Comment by fkleuver** [comment]
The only difference between `@renderStrategy` and `@bindingCommand` is in the storage mechanism; the binding commands are registered and resolved exclusively via DI, whereas render strategy is registered with DI, resolved once, and then stored on the prototype. This is for performance because the renderer is such a hot path.
Are you looking for the flexibility of having different sets of render strategies for different component contexts? That's the only difference there is atm from a functional point of view.

**Comment by bigopon** [comment]
The renderer contract with rendering engine at the moment is quite simple: only a render method that takes a renderable, targets, definition, host and replacement parts. From that joint, we can probably break the renderer out of runtime and let it sit in its own package, or merge with JIT. Doing so we have a plug-able renderer, and it could make more sense for renderer and compiler to sit together in same package. Then registering different targeted instruction type will be made configurable, unlike what we have at the moment, always assume there is a fixed set of core instruction type

**Comment by EisenbergEffect** [comment]
I don't think it can move to the JIT because we need it at runtime for AOT.

**Comment by bigopon** [comment]
When we move it out of runtime, there is nothing stopping us from having different renderers for different target. JIT, AOT, each of them can have their own renderer, or they can share the same renderer. say AOT targeting web will share the same renderer with JIT-web etc

**Comment by EisenbergEffect** [comment]
I thought you were saying to put it in the JIT. It could be moved to its own package, yes. I'm a bit wary of having lots of packages though. Maybe if someone wanted to do a NativeScript renderer in its own package to prove out the idea...hint...hint...

**Comment by fkleuver** [comment]
I think there's a broader discussion to be had here in that case. There's multiple places where we can make this decision:
- ObserverLocator + observers (tie observer location strategies to observers, make them pluggable and tree shakeable)
- Binding commands (if you want to include the jit in the app but not necessarily have all commands)
- Render strategies (what this issue is about)
- Bindings (binding, call, listener, ref) which are kind of tied to render strategies

I do like the idea of splitting infrastructure from implementations, even for core functionality. If anything, dogfeeding our extensibility points for the core will improve their quality and act as proof that it works well.

**Comment by bigopon** [comment]
> Render strategies (what this issue is about)

If it didn't mean what you mentioned in `Is this different from the @renderStrategy decorator I added a while ago?`, then it would be the right thing. I guess we can create a `IRenderer` from `DI.createInterface`, so renderer can be registered depending on the scenario, just like other interfaces. It will also require re-structuring of the observer locator like Fred said as well, as different renderers may employ different 
observation determination strategies

**Comment by fkleuver** [comment]
>  as different renderers may employ different observation determination strategies

Afaik the renderer just glues together nodes, instructions and resources. I don't think it should even know about observers, let alone deal with observation strategies.

Could you give an example where it might be beneficial to have different renderer implementations for the same kind of instruction?

**Comment by bigopon** [comment]
It's a chain effect of the renderer. For me, the graph looks like this:

```
instructions -> Renderer -> components / bindings -> observer-locator -> observers
```
There is nothing much in the rest of the runtime package that need to know about this chain. A renderer only works with a certain set of bindings, and bindings employ observer locator. Different platform target may require partially different observer locator configuration.

Hope it makes sense why your idea of externally configurable observer locator is a very nice API to have.

**Comment by fkleuver** [comment]
> Different platform target may require partially different observer locator configuration.

> For example, hydrating element in a DOM environment will be different to hydrating element in a web VR environment, or a canvas based renderer.

It makes perfect sense, but if support for non-DOM environments is the primary motivator then imo this can wait. It's easy to refactor out of platform specific stuff - easier than to try and design ahead for it (unless you've done it several times before and already know the recipes that work)

Let's have everything working and fully tested before we start layering on abstractions that we won't need on the short term. We may very well be able to generate other variants via AOT as well, which will be more performant anyway.

**Comment by fkleuver** [comment]
I'd just like to add to this that I'm a firm believer in having more than one implementation *before* creating an abstraction, because that's when they naturally emerge and tend to make more sense. Having two implementations is better than one, and having three is better still.

What this translates to is that we'd implement a fully standalone WebGL renderer first, then put the two renderers next to each other and extract out an abstraction that actually scratches the itch. That will be more work than "just a matter of implementing a couple of interfaces", sure, but it's deferring the work to when we have more information to do that work. Maybe you're worried about it being more work the later you do it, which is true in terms of how many LOC need to change, but it will be easier too.

Long story short: @bigopon maybe can you implement such a renderer? That would give me a better idea of what it's going to as well. I've never really worked with non-DOM environments so I wouldn't know how to support you guys in creating these abstractions without knowing the implementations

**Comment by fkleuver** [comment]
Fixed by #281
